### PR TITLE
Visual upgrade + code changes

### PR DESCRIPTION
**Pull request by b7g**
# Description
This PR includes a visual and a code makeover.
## Visual
splash screen
n1

new menu
fix1

new info dialog
fix2

new section about/contribute/report a bug
fix3

new section settings
fix4

the new ingame pause menu
n6

## Changes, changes everywhere
Additionally to the visual changes depicted above following changes were made:
- added new Utils Singleton (for commonly used utilities)
- function handle_error was moved from GameManager -> Utils
- added new UserSettings singleton (loads, saves and handles the user settings)
- overhauled the GameManager singleton quite a bit
- added new PlayerManager singleton (split from old GameManager)
- bbcode support for game descriptions was removed for good
- pause_menu got an overhaul as well
- translations for the main app were added, translations for the games and descriptions still need to be done.

## Type of change
- [x] Bug fix (non breaking change that fixes an issue)
- [x] New feature (non breaking change that adds functionality)
- [x] Tooling (non breaking change that adds functionality to the workflow)
- --- Breaking change (fix or feature that causes existing functionality to not work as expected)

## Environment
- OS: Ubuntu 22.04 LTS
- Software versions: Godot 3.4

## Testing
tested everything at some point, should work

# Checklist
- [x] My code follows the general style guidelines of the project
- [x] I have perfomed a self-review of my code
- [x] I have commented my code, particulary where it is unclear
- [x] My changes generate no new warnings or errors
- [x] The project compiles and runs correctly

### CODE REVIEW

**Review by RedstoneMedia — CHANGES_REQUESTED** [review]
Why would you need an image of a black/white square png image. Just use a color rect, if you need the spacing, just use containers (margin) for that.

**Comment by RedstoneMedia** [comment]
Also, please capitalize the first character in a sentence and words on buttons (example, "spielen" instead of "Spielen")

**Comment by Numenter** [comment]
Windows error:
data_manager.gd:41 -> error ERR_CANT_CREATE -> "user://savefiles/p/res://games/flappybird/"

Why is game_id a path? "res://games/flappybird"

**Comment by RedstoneMedia** [comment]
Also `_last_played` in `game_card.gd` cannot be of type int, since  `get_time_last_played` can return null, which causes an immediate crash, if the game has not been played yet. 

Using a static video instead of animations in Godot is also a problem, on larger monitors and not that easy to edit.

Another thing to note is, that not every game has a high score, so it should not be displayed in the game info.

**Comment by Joshix-1** [comment]
0 shouldn't be the default for score in end_game. There should be the option for games without scores. (null as default)

**Comment by Numenter** [comment]
Windows error:
WebM"s stutter on windows and the support for WebM is removed in 4.0
https://github.com/godotengine/godot/issues/55815
Theora is recommended

**Comment by Numenter** [comment]
> Why would you need an image of a black/white square png image. Just use a color rect, if you need the spacing, just use containers (margin) for that.

The black square imange is used in Settings/Boot Splash, which has to be an image.
The white square imange is used in the blur shader.

**Comment by Numenter** [comment]
Not a fan of the cryptic translation codes.
image
image

T_PAUSED, T_RESUME, T_RESTART, T_BACK_TO_MENU would be a better fit.

PS: here is a PR for this https://github.com/b7g/Suffragium/pull/1

**Comment by Joshix-1** [comment]
Yes translation codes should always contain the whole english text, then you know when you have to update all the other langauge variations

File: game/app/scenes/game_card.gd [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+onready var _popup_label_playtime := $PopupDialogInfo/VC/PC/MC/HC/MC/VC/HC/LabelPlaytimeNumber
+onready var _popup_label_playtime_unit := $PopupDialogInfo/VC/PC/MC/HC/MC/VC/HC/LabelPlaytimeUnit
+onready var _popup_label_highscore := $PopupDialogInfo/VC/PC/MC/HC/VC/LabelHighscore
+
+
+func setup(game_config: ConfigFile):
+	_game_config = game_config
+
+	var game_folder_path = game_config.get_meta("folder_path")
+	_playtime = GameManager.get_playtime(game_folder_path)
```
Comment (Joshix-1) [inline game/app/scenes/game_card.gd]: The key should be the folder_name not the whole path

File: game/app/scenes/game_card.gd [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	_last_played = GameManager.get_time_last_played(game_folder_path)
+	var highscore_dict = GameManager.get_highscore(game_folder_path)
+	var highscore = "-"
+	if highscore_dict.has("score"):
+		highscore = str(highscore_dict["score"])
+	_popup_label_highscore.text = highscore
```
Comment (Joshix-1) [inline game/app/scenes/game_card.gd]: Only display highscore if one is present. There are games that don't have highscores

File: game/app/scenes/game_card.gd [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+func get_playtime() -> float:
+	return _playtime
+
+
+func get_last_played() -> int:
+	return _last_played
+
+
+func _format_playtime(playtime) -> Dictionary:
```
Comment (Joshix-1) [inline game/app/scenes/game_card.gd]: I don't like how inaccurate this is

File: game/app/singleton/game_manager.gd [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	var main_scene = game_config.get_value("game", "main_scene")
+	var scene_path = "%s/%s" % [game_folder, main_scene]
+	var err = _change_scene(scene_path)
+	if err == OK:
+		_game_started(game_folder, game_config)
+		_pause_menu = res_pause_menu.instance()
+		get_tree().get_root().add_child(_pause_menu)
+
+
+func end_game(message: String = "", score: int = 0):
```
Comment (Joshix-1) [inline game/app/singleton/game_manager.gd]: Score should be per default null, to make a difference between no score and a score of 0.

File: game/app/singleton/game_manager.gd [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+
+# load a config file into _games
+func _load_game_config_file(game_config_path: String):
+	var config_file = ConfigFile.new()
+	var err = config_file.load(game_config_path)
+	if err:
+		push_error("Could not load game config at path '%s'" % game_config_path)
+		return
+	config_file.set_meta("folder_path", game_config_path.get_base_dir())
```
Comment (Joshix-1) [inline game/app/singleton/game_manager.gd]: Better save the folder_name, so it can be used as key later

File: game/app/singleton/game_manager.gd [hunk 6]
```diff
@@ -0,0 +1,157 @@
+extends Node
+
+var _menu_path = "res://app/scenes/menu.tscn"
+var _games_folder_path = "res://games"
+var _games = []
```
Comment (Joshix-1) [inline game/app/singleton/game_manager.gd]: why an array? I think the dictionary solution was better

File: game/app/singleton/player_manager.gd [hunk 7]
```diff
@@ -0,0 +1,5 @@
+extends Node
+
+
+func get_current_player() -> String:
```
Comment (Joshix-1) [inline game/app/singleton/player_manager.gd:4]: Why move that here, if it still has no functionality?

File: game/menu/README.md [hunk 8]
```diff
@@ -3,7 +3,7 @@
 ## Game-Id
 If in the code a `game_id` is mentioned, this is the name of the folder where the game is in.
 
-## GameManager
+## GameManger
```
Comment (Joshix-1) [inline game/menu/README.md]: Why? GameManger?

File: game/translations/tr.csv [hunk 9]
```diff
@@ -0,0 +1,43 @@
+,de,en
```
Comment (Joshix-1) [inline game/translations/tr.csv]: I would use the english text as key and have one file for every language

**Review by Joshix-1 — CHANGES_REQUESTED** [review]
Just looked over the code and noticed some problems

File: game/translations/tr.csv (hunk 9 above)
Comment (Numenter) [reply game/translations/tr.csv]: Do you mean "About Suffragium" or "T_ABOUT_SUFFRAGIUM" as key.

This is how I would do it. https://github.com/b7g/Suffragium/pull/1
```
T_ABOUT_SUFFRAGIUM,Über Suffragium,About Suffragium
```
And maybe switch EN to the front and remove "T_".

**Comment by b7g** [comment]
Thanks for the lots of comments and info, I appreciate it.
Seems like shit hit the fan a bit haha.

I will respond to the comments first and then resolve the issues.

**Comment by b7g** [comment]
> Also, please capitalize the first character in a sentence and words on buttons (example, "spielen" instead of "Spielen")

I think you meant it the other way around in your example?

**Comment by b7g** [comment]
> Windows error: data_manager.gd:41 -> error ERR_CANT_CREATE -> "user://savefiles/p/res://games/flappybird/"
> 
> Why is game_id a path? "res://games/flappybird"

shouldn't be, you are right
I'm a little confused there I have to say, as I haven't really touched the data_manager at all.

**Comment by b7g** [comment]
> Also `_last_played` in `game_card.gd` cannot be of type int, since `get_time_last_played` can return null, which causes an immediate crash, if the game has not been played yet.
> 
> Using a static video instead of animations in Godot is also a problem, on larger monitors and not that easy to edit.
> 
> Another thing to note is, that not every game has a high score, so it should not be displayed in the game info.

a. correct
b. I don't see how an more complex animation can be easily recreated with code. Also other games use videos too, so there should be at least a way to prevent the issues, if they are present right now.
c. just for clarification: just the games that have no score are meant?

**Comment by b7g** [comment]
> 0 shouldn't be the default for score in end_game. There should be the option for games without scores. (null as default)

I agree

**Comment by b7g** [comment]
> Windows error: WebM"s stutter on windows and the support for WebM is removed in 4.0 [godotengine/godot#55815](https://github.com/godotengine/godot/issues/55815) Theora is recommended

Will be changed

**Comment by b7g** [comment]
> Not a fan of the cryptic translation codes.
> 
> T_PAUSED, T_RESUME, T_RESTART, T_BACK_TO_MENU would be a better fit.
> 
> PS: here is a PR for this [b7g#1](https://github.com/b7g/Suffragium/pull/1)

I agree on this, but had problems with the sizing in the past.
- Godot did not allow smaller buttons then ones the translation codes fit into. Maybe that is fixed by now or we can work around that by always using containers.

File: game/app/scenes/game_card.gd (hunk 1 above)
Comment (b7g) [reply game/app/scenes/game_card.gd]: you are right

File: game/app/scenes/game_card.gd (hunk 2 above)
Comment (b7g) [reply game/app/scenes/game_card.gd]: will change

File: game/app/scenes/game_card.gd (hunk 3 above)
Comment (b7g) [reply game/app/scenes/game_card.gd]: I think it is an improvement
Before
- it was always displayed in seconds
- it had 3 decimal places

The target audience of this application are not measurement engineers right?

**Comment by RedstoneMedia** [comment]
> The black square imange is used in Settings/Boot Splash, which has to be an image.

I'm pretty sure there is an option to set a custom boot splash screen instead of just an image. If that is not the case just make the image a 1x1 transparent image and make the background black ( this would also only appear very shortly before a splash screen scene is loaded ). 

> The white square imange is used in the blur shader.

And you can't create a rectangle in a shader ?

File: game/app/singleton/game_manager.gd (hunk 4 above)
Comment (b7g) [reply game/app/singleton/game_manager.gd]: I agree

**Comment by RedstoneMedia** [comment]
> > Also, please capitalize the first character in a sentence and words on buttons (example, "spielen" instead of "Spielen")
> 
> I think you meant it the other way around in your example?

Yeah sorry my bad. The other way around is obviously correct.

File: game/app/singleton/game_manager.gd (hunk 5 above)
Comment (b7g) [reply game/app/singleton/game_manager.gd]: I agree

File: game/app/singleton/game_manager.gd (hunk 6 above)
Comment (b7g) [reply game/app/singleton/game_manager.gd]: Good question, I agree

File: game/app/singleton/player_manager.gd (hunk 7 above)
Comment (b7g) [reply game/app/singleton/player_manager.gd:4]: a. why should the GameManager handle the players, even if it is not really functional yet?
b. de-clutters the GameManager

File: game/menu/README.md (hunk 8 above)
Comment (b7g) [reply game/menu/README.md]: type error, will change

**Comment by RedstoneMedia** [comment]
> c. just for clarification: just the games that have no score are meant?

Ideally yes. However this would require another manual entry in the game CFGs. That's why hiding the highscore for all games until a non null highscore is returned by end_game, is also a good option.

**Comment by Numenter** [comment]
> > The black square imange is used in Settings/Boot Splash, which has to be an image.
> 
> I'm pretty sure there is an option to set a custom boot splash screen instead of just an image. If that is not the case just make the image a 1x1 transparent image and make the background black ( this would also only appear very shortly before a splash screen scene is loaded ).

A 1x1 transparent image works fine for the boot splash and it can be used for the blur shader too.

> > The white square imange is used in the blur shader.
> 
> And you can't create a rectangle in a shader ?

Maybe, I'm not a shader guy.

**Comment by b7g** [comment]
> I'm pretty sure there is an option to set a custom boot splash screen instead of just an image. If that is not the case just make the image a 1x1 transparent image and make the background black ( this would also only appear very shortly before a splash screen scene is loaded ).

There is not. Will change to pixel.

> And you can't create a rectangle in a shader ?

I just did it like in the example project provided by Godot.
Maybe that is possible, but my knowledge with shaders is lacking.

File: game/translations/tr.csv (hunk 9 above)
Comment (b7g) [reply game/translations/tr.csv]: I merged the pr, so the keys are fixed.
But I don't think multiple files are a good idea, it just makes it more work to sync them.

**Comment by RedstoneMedia** [comment]
> Maybe that is possible, but my knowledge with shaders is lacking.

I did it without the image file. Turns out all you have to do is change the TextureRect to a ColorRect. The color of the rect doesn't matter since its overwritten anyway by the shader.

The PR: https://github.com/b7g/Suffragium/pull/3

**Comment by Numenter** [comment]
Simplified the Tween And Timer Horror. :P https://github.com/b7g/Suffragium/pull/4

**Comment by ASecondGuy** [comment]
The menu sene hurts to look at. I'll make a PR to fix that mess.
Why did you remove bbcode support? That is super easy to implement and gives so many options. That doesn't make any sense.
Also. All those fancy buttons are bad practice. You should use one Button node and change the look with a theme. (also in my future PR)

**Comment by RedstoneMedia** [comment]
> Also. All those fancy buttons are bad practice. You should use one Button node and change the look with a theme. (also in my future PR)

Ah do your also working on that ? How far into the process are you with creating a theme for the buttons and stuff ? Because I want to know if it's worth it to continue work on my PR. (I would not fix the other issues you mentioned)

**Comment by ASecondGuy** [comment]
> > Also. All those fancy buttons are bad practice. You should use one Button node and change the look with a theme. (also in my future PR)
> 
> Ah do your also working on that ? How far into the process are you with creating a theme for the buttons and stuff ? Because I want to know if it's worth it to continue work on my PR. (I would not fix the other issues you mentioned)

I'm just starting. What exactly are you fixing so I can avoid that? 
And if we both change nodes or resources in the same scene it will probably give merge requests so we should watch out for that.
@RedstoneMedia

**Comment by RedstoneMedia** [comment]
> I'm just starting. What exactly are you fixing so I can avoid that? 
> And if we both change nodes or resources in the same scene it will probably give merge requests so we should watch out for that.
> @RedstoneMedia 

I'm creating a theme for the buttons, dorp-downs, check boxes, labels, etc. Because overriding the styles on every single button and label, as it is currently is just painful to look at. So I would probably touch allot of UI nodes, that have a theme override. I won't change the tree structure of the nodes though.

**Comment by ASecondGuy** [comment]
Ok. Then I can change the Scenes all I want and use your theme once it is merged.

**Comment by RedstoneMedia** [comment]
> Ok. Then I can change the Scenes all I want and use your theme once it is merged.

Yup, that will probably work.

**Comment by RedstoneMedia** [comment]
Ok, so I was about to change the buttons, but I realized that they are made out of more than one node. The actual button object isn't even really styled. That means that I will need to remove those extra nodes from the tree, to style them.

**Comment by b7g** [comment]
> The menu sene hurts to look at. I'll make a PR to fix that mess. Why did you remove bbcode support? That is super easy to implement and gives so many options. That doesn't make any sense.

Because bbcode in the menu “hurts to look at”. Something should not be done because it’s possible, but because it makes sense. And bbcode in the menu does not make sense, as it gives unnecessary freedoms:
- to break the readability,
- to make it look inconsistent

> Also. All those fancy buttons are bad practice. You should use one Button node and change the look with a theme. (also in my future PR)

I use Godot to it’s potential with my abilities. And it makes it really hard to properly style things in my opinion. If you are able to do it better, go for it. I would like to see.

**Comment by ASecondGuy** [comment]
~~BBcode is the best thing ever~~ You can do a lot of ~~Stuff~~ nice Texteffects to make certain parts **stand out**.
It is not just possible but gives _**possibilities**_.
Of course you can use it to break readability and make it look inconsistent but you can use any tool wrong.
For example you used the Button Node wrong and made the Scenes not readable.
You used Themes wrong and that makes accidental inconsistency way more likely. Especially when the look is changed in the future.

When someone introduces inconsistency you help them fix it. You don't remove options in case someone uses them incorrectly. 
But that last paragraph is just my opinion.

The themes are actually very easy to use once you understand them. But you don't have to do that if you don't want. I'm already working on fixing all that.

**Comment by b7g** [comment]
When you need to highlight something in the description, it just means you haven’t done your job properly and are bringing it together with highlights in the description on the cost of the user.

**Comment by ASecondGuy** [comment]
> When you need to highlight something in the description, it just means you haven’t done your job properly and are bringing it together with highlights in the description on the cost of the user.

The Description isn't to bandaid over flaws of the game, I agree. If someone uses it like that you should tell them to stop and fix the game.
The description is to describe the game (mind bending concept I know). That description can sometimes be improved with the options bbcode offers. I see no downsides to including it. (Again, someone misusing a tool is not a reason to remove the tool)

**Comment by b7g** [comment]
Please note that I swapped out the images in the description to reflect the latest changes.
The changes are rather subtle, therefore the heads-up.

**Comment by Numenter** [comment]
Bug:
Crash on Soriting : Last Played when not all games have been played

Reproduce:
1.  Remove savegames
2. Sort by last played

ERROR menu.gd:144 Invalid operands 'Nil' and 'Nil' in operator '>'.

Could be the same bug @RedstoneMedia stated before.
> Also `_last_played` in `game_card.gd` cannot be of type int, since `get_time_last_played` can return null, which causes an immediate crash, if the game has not been played yet.

**Comment by b7g** [comment]
should be finally fixed :')

**Comment by b7g** [comment]
@Joshix-1

I made the formatting of the playtime more precise.

- now also seconds are displayed if minutes were not reached yet
- when the playtime is between 1 - 10 hours also the minutes are displayed in the popup dialog

maybe that's precise enough for your liking?

**Comment by ASecondGuy** [comment]
I finished the scene tree refactor of this. 
Just need to configure the theme so it looks like it did before.

**Comment by Joshix-1** [comment]
Sounds better. Could we maybe show even more precise times on hover? x days y hours z minutes

**Comment by RedstoneMedia** [comment]
> Just need to configure the theme so it looks like it did before.

You mean how it looks now right ? You don't want to change the look back to before my (and b7g's) improvements. Shouldn't you be able to just apply the theme (instead of changing it) ?

A button is going to stay a button, how would a scene tree refactor, change this ?

**Comment by ASecondGuy** [comment]
> > Just need to configure the theme so it looks like it did before.
> 
> You mean how it looks now right ? You don't want to change the look back to before my (and b7g's) improvements. Shouldn't you be able to just apply the theme (instead of changing it) ?
> 
> A button is going to stay a button, how would a scene tree refactor, change this ?

I improved the scene tree structure and removed unnecessary nodes.
Because there were still theme overwrites used and the theme didn't have all nodes or they were configured differently it looks like crap.
I'm now updating the theme to get the intended look. (I'm not gonna search through old files to find the exact numbers and colors)
I'm not much of an artist so someone probably needs to go over the colors again.
The buttons stated exactly the same btw.

**Comment by b7g** [comment]
> Sounds better. Could we maybe show even more precise times on hover? x days y hours z minutes

It’s common for playtime to be displayed in hours, not minutes or days

**Comment by RedstoneMedia** [comment]
@ASecondGuy 
> I'm not gonna search through old files to find the exact numbers and colors

Okay ? That's a bit suspicious.

Edit: From what I can see, only the top bar looks really off right now. That could be handled with overrides (Or a new custom theme just for the menu bar)

> Because there were still theme overwrites used

Also theme overrides are not the devil (If its only used once or twice)

Anyways, your current branch has some obvious issues (I know WIP and stuff but still):
- `About.gd` needs to be renamed to `about.gd`
- The label assertion in `About.gd` always fails on my system (Not just in weird instances. It can **not** be safely ignored)

**Comment by ASecondGuy** [comment]
@RedstoneMedia What is suspicious about that? Most of the visuals were in the theme overwrites before.
I reorganized the nodes and therefore most visuals went missing. I don't wanna search for them so I approximated.
Btw. the menu bar is a PanelContainer and gets its visuals directly from the main_theme.

And I know overwrites aren't inherently bad. But I found at least a handful of them in V/HBoxContainers where they make limited sense. And I removed many of those Containers.

File: game/project.godot [hunk 10]
```diff
@@ -29,13 +24,12 @@ _global_script_classes=[ {
 "language": "GDScript",
 "path": "res://games/sortit/player.gd"
 }, {
-"base": "MarginContainer",
+"base": "Reference",
```
Comment (Numenter) [inline game/project.godot:27]: sort_it.gd extends form MarginContainer and my godot is changing this back to MarginContainer.

File: game/app/scenes/data_manager.gd [hunk 11]
```diff
@@ -86,7 +86,7 @@ class ScoreSorter:
 
 
 ## Handle data for the game that ended.
-func game_ended(player: String, game_id: String, start_time, score: Dictionary):
+func game_ended(player: String, game_id: String, start_time, score_or_null):
```
Comment (Numenter) [inline game/app/scenes/data_manager.gd:89]: Don't like the name. Is it score or is it null or is the score null?
Just call it score and add a comment with which values are expected.

**Review by Numenter — APPROVED** [review]
Just two little things, but you got my :+1: anyway.

File: game/project.godot (hunk 10 above)
Comment (RedstoneMedia) [reply game/project.godot:27]: Yeah, this is weird, my godot always changes it back and forth. I honestly have no Idea what causes this.

**Review by ASecondGuy — APPROVED** [review]
Now this is a good PR I can get behind.

**Review by RedstoneMedia — APPROVED** [review]

**Pull request merged**